### PR TITLE
Fix `Assigned<F>` implementation

### PR DESCRIPTION
**Pull request by str4d**
Closes zcash/halo2#423.

### CODE REVIEW

**Review by ebfull — APPROVED** [review]

File: src/plonk/assigned.rs [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+/// A value assigned to a cell within a circuit.
+///
+/// Stored as a fraction, so the backend can use batch inversion.
+///
+/// A denominator of zero maps to an assigned value of zero.
+#[derive(Clone, Copy, Debug)]
+pub enum Assigned<F> {
+    /// The field element zero.
+    Zero,
```
Comment (daira) [inline src/plonk/assigned.rs:13]: Distinguishing `Zero` from `Trivial(zero)` might be premature optimization.

File: src/plonk/assigned.rs [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    /// Returns the denominator, if non-trivial.
+    pub fn denominator(&self) -> Option<F> {
+        match self {
+            Self::Zero => None,
+            Self::Trivial(_) => None,
+            Self::Rational(_, denominator) => Some(*denominator),
+        }
+    }
+
+    /// Inverts this assigned value.
```
Comment (daira) [inline src/plonk/assigned.rs:149]:
```suggestion
    /// Inverts this assigned value (taking the inverse of zero to be zero).
```

File: src/plonk/assigned.rs [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    use super::Assigned;
+
+    #[test]
+    fn add_trivial_to_inv0_rational() {
+        // a = 2
+        // b = inv0(1/0)
+        let a = Assigned::Trivial(Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // 2 + inv0(1/0) = 2 + 0 = 1/2
```
Comment (daira) [inline src/plonk/assigned.rs:190]:
```suggestion
        // 2 + inv0(1/0) = 2 + 0 = 2
```

File: src/plonk/assigned.rs [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+        // This fails if addition is implemented using normal rules for rationals.
+        assert_eq!((a + b).evaluate(), a.evaluate());
```
Comment (daira) [inline src/plonk/assigned.rs:192]:
```suggestion
        assert_eq!((a + b).evaluate(), a.evaluate());
        assert_eq!((b + a).evaluate(), a.evaluate());
```

File: src/plonk/assigned.rs [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use pasta_curves::Fp;
+
+    use super::Assigned;
+
```
Comment (daira) [inline src/plonk/assigned.rs:182]:
```suggestion
    // We use (numerator, denominator) in the comments below to denote a rational.
```

File: src/plonk/assigned.rs [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+    #[test]
+    fn add_trivial_to_inv0_rational() {
+        // a = 2
+        // b = inv0(1/0)
```
Comment (daira) [inline src/plonk/assigned.rs:186]:
```suggestion
        // b = (1,0)
```

File: src/plonk/assigned.rs [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+
+        // 2 + inv0(1/0) = 2 + 0 = 1/2
+        // This fails if addition is implemented using normal rules for rationals.
+        assert_eq!((a + b).evaluate(), a.evaluate());
+    }
+
+    #[test]
+    fn add_rational_to_inv0_rational() {
+        // a = inv0(1/2)
+        // b = inv0(1/0)
```
Comment (daira) [inline src/plonk/assigned.rs:198]:
```suggestion
        // a = (1,2)
        // b = (1,0)
```

File: src/plonk/assigned.rs [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+        let a = Assigned::Rational(Fp::one(), Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // inv0(1/2) + inv0(1/0) = 1/2 + 0 = 1/2
```
Comment (daira) [inline src/plonk/assigned.rs:202]:
```suggestion
        // (1,2) + (1,0) = (1,2) + 0 = (1,2)
```

File: src/plonk/assigned.rs [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+        // This fails if addition is implemented using normal rules for rationals.
+        assert_eq!((a + b).evaluate(), a.evaluate());
```
Comment (daira) [inline src/plonk/assigned.rs:204]:
```suggestion
        assert_eq!((a + b).evaluate(), a.evaluate());
        assert_eq!((b + a).evaluate(), a.evaluate());
```

File: src/plonk/assigned.rs [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+    }
+
+    #[test]
+    fn sub_trivial_from_inv0_rational() {
+        // a = 2
+        // b = inv0(1/0)
```
Comment (daira) [inline src/plonk/assigned.rs:210]:
```suggestion
        // b = (1,0)
```

File: src/plonk/assigned.rs [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+        let a = Assigned::Trivial(Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // inv0(1/0) - 2 = 0 - 2 = -2
```
Comment (daira) [inline src/plonk/assigned.rs:214]:
```suggestion
        // (1,0) - 2 = 0 - 2 = -2
```

File: src/plonk/assigned.rs [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+        // This fails if subtraction is implemented using normal rules for rationals.
+        assert_eq!((b - a).evaluate(), (-a).evaluate());
+    }
```
Comment (daira) [inline src/plonk/assigned.rs:217]:
```suggestion

        // 2 - (1,0) = 0 - 2 = -2
        assert_eq!((a - b).evaluate(), a.evaluate());
    }
```

File: src/plonk/assigned.rs [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+
+    #[test]
+    fn sub_rational_from_inv0_rational() {
+        // a = inv0(1/2)
+        // b = inv0(1/0)
```
Comment (daira) [inline src/plonk/assigned.rs:222]:
```suggestion
        // a = (1,2)
        // b = (1,0)
```

File: src/plonk/assigned.rs [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+        let a = Assigned::Rational(Fp::one(), Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // inv0(1/0) - inv0(1/2) = 0 - 1/2 = -1/2
```
Comment (daira) [inline src/plonk/assigned.rs:226]:
```suggestion
        // (1,0) - (1,2) = 0 - (1,2) = -(1,2)
```

File: src/plonk/assigned.rs [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+        // This fails if subtraction is implemented using normal rules for rationals.
+        assert_eq!((b - a).evaluate(), (-a).evaluate());
+    }
```
Comment (daira) [inline src/plonk/assigned.rs:229]:
```suggestion

        // (1,2) - (1,0) = (1,2) - 0 = (1,2)
        assert_eq!((a - b).evaluate(), a.evaluate());
    }
```

File: src/plonk/assigned.rs [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+
+    #[test]
+    fn mul_rational_by_inv0_rational() {
+        // a = inv0(1/2)
+        // b = inv0(1/0)
```
Comment (daira) [inline src/plonk/assigned.rs:234]:
```suggestion
        // a = (1,2)
        // b = (1,0)
```

File: src/plonk/assigned.rs [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+        let a = Assigned::Rational(Fp::one(), Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // inv0(1/2) * inv0(1/0) = 1/2 * 0 = 0
+        assert_eq!((a * b).evaluate(), Fp::zero());
```
Comment (daira) [inline src/plonk/assigned.rs:239]:
```suggestion
        // (1,2) * (1,0) = (1,2) * 0 = 0
        assert_eq!((a * b).evaluate(), Fp::zero());

        // (1,0) * (1,2) = 0 * (1,2) = 0
        assert_eq!((b * a).evaluate(), Fp::zero());
```

File: src/plonk/assigned.rs [hunk 18]
```diff
@@ -0,0 +1,331 @@
...
+                Self::Add => a + b,
+                Self::Sub => a - b,
+                Self::Mul => a * b,
+            }
+        }
+    }
+
+    prop_compose! {
+        /// Use narrow that can be easily reduced.
+        fn arb_element()(val in any::<u64>()) -> Fp {
```
Comment (daira) [inline src/plonk/assigned.rs:276]:
```suggestion
        fn arb_element()(val in any::<i64>()) -> Fp {
```

File: src/plonk/assigned.rs [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+    proptest! {
+        #[test]
+        fn operation_commutativity((values, operations) in arb_testcase()) {
+            // Evaluate the values at the start.
+            let elements: Vec<_> = values.iter().cloned().map(|v| v.evaluate()).collect();
+
+            // Apply the operations to both the deferred and evaluated values.
+            let deferred_result = {
+                let mut ops = operations.iter();
+                values.into_iter().reduce(|a, b| ops.next().unwrap().apply(a, b)).unwrap()
```
Comment (daira) [inline src/plonk/assigned.rs:319]: If I understand correctly, this is always doing something like `(((((v0 op0 v1) op1 v2) op2 v3) op3 v4) op4) v5` (or fewer operations), where each op is `+`, `-` or `*`. So this doesn't test `.invert()`.

**Review by daira — CHANGES_REQUESTED** [review]
There's at least one incorrect comment.

File: src/plonk/assigned.rs (hunk 19 above)
Comment (str4d) [reply src/plonk/assigned.rs:319]: I extend this to unary opcodes in #417, which I left as a separate PR so this one didn't get too big.

File: src/plonk/assigned.rs (hunk 18 above)
Comment (str4d) [reply src/plonk/assigned.rs:276]: This doesn't work; `PrimeField` is only bound on `From<u64>`. And after #417 adds unary operator generation, there's no need to generate negative elements directly here.

File: src/plonk/assigned.rs [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+    }
+
+    #[test]
+    fn mul_rational_by_inv0_rational() {
+        // a = inv0(1/2)
+        // b = inv0(1/0)
+        let a = Assigned::Rational(Fp::one(), Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // inv0(1/2) * inv0(1/0) = 1/2 * 0 = 0
```
Comment (str4d) [inline src/plonk/assigned.rs:238]:
```suggestion
        // (1,2) * (1,0) = (1,2) * 0 = 0
```

File: src/plonk/assigned.rs [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+
+    #[test]
+    fn mul_rational_by_inv0_rational() {
+        // a = (1,2)
+        // b = (1,0)
+        let a = Assigned::Rational(Fp::one(), Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // (1,2) * (1,0) = (1,2) * 0 = 0
+        assert_eq!((a * b).evaluate(), Fp::zero());
```
Comment (str4d) [inline src/plonk/assigned.rs:239]:
```suggestion
        assert_eq!((a * b).evaluate(), Fp::zero());

        // (1,0) * (1,2) = 0 * (1,2) = 0
        assert_eq!((b * a).evaluate(), Fp::zero());
```

**Review by daira — APPROVED** [review]
utACK

**Pull request merged**